ecgtools/core.py

```diff
@@ -404,7 +421,10 @@ def _parse_file_attributes(filepath: str, parser: callable = None):
 
     results = {'path': filepath}
     if parser is not None:
-        x = parser(filepath)
+        if len(local_attrs.keys()) == 0:
+            x = parser(filepath)
+        else:
+            x = parser(filepath, local_attrs)
```

Comment:
If `local_attrs` defaults to `{}`, then this block could just be:

```suggestion
        x = parser(filepath, local_attrs)
```

Reply:
@kmpaul this change is a little trickier.  The reason being is currently the `parser` function points to either `_parser_netcdf` or `cmip6_default_parser`.  `_parser_netcdf` s second argument is a `dictionary`, where as `cmip6_default_parser` can have a `list` as a second argument.   So either way, this code is incorrect and will need to be modified.  Thanks for catching this and I'll have to think about how this should be addressed better.

Reply:
Ok.  Then the appropriate way of handling this in Python is to unify the `parser` call signature.  That is, the generic `parser` signature should be changed like so:

```python
def parser(common_arg1, common_arg2, ..., **kwargs):
```

All arguments that are common to all parsers should be listed upfront, and any other arguments that are specific to the kind of parser should be passed in as keyword arguments via `kwargs`.  Therefore, any arguments (e.g., `local_attrs=my_local_attrs`) that aren't common to all parsers can be retrieved in the specific parser's function/init like `kwargs['local_attrs']`.  This will let you *always* pass in `local_attrs`, even if the other parsers don't use it.

Then, this line can be called with:

`x = parser(filepath, local_attrs=local_attrs)`
Then, you know that in your use cases, you can always call the

Reply:
My original plan in regards to the `parser` callable was to only expose the `filepath` parameter only i.e. `parser(filepath)`. For a user defined parser with multiple parameters, the user would then need to use `functools.partial()` to create a partial function. As an example the current default cmip6 parser signature looks like this

```python
def cmip6_default_parser(
    filepath: str,
    global_attrs: list,
    variable_attrs: list = None,
    attrs_mapping: dict = None,
    add_dim: bool = True,
):
```

However, you can't use it as is in the `Builder()` class. The user needs to create a partial function in which the other arguments other than filepath are provided. 

```python

cmip6_parser = functools.partial(
    cmip6_default_parser,
    global_attrs=cmip6_global_attrs,
    variable_attrs=cmip6_variable_attrs,
    attrs_mapping=cmip6_attrs_mapping,
)
```


This is then used as follows:

```pythn
b = Builder(
        root_path,
        depth=depth,
        extension='*.nc',
        exclude_patterns=['*/files/*', '*/latest/*'],
        lazy=lazy,
        parser=cmip6_parser,
    )
```

So, to keep the existing API without modifying it, we could create a partial function from `self._parser_netcdf`:

```python
partial_parser = functools.partial(self._parser_netcdf, local_attrs=local_attrs) # Needs to be tested
self.builder = Builder(None, parser=partial_parser, lazy=False)
```

Any thoughts? Does my explanation even make sense? 😀

Reply:
@andersy005:  Yes.  `functools.partial` is another way of doing this.  I think that is clear enough.

Reply:
This generic interface allows people to have different `local_attrs` for different datasets.  Creating a `partial_parser` would require a new `Builder` class to be created per `local_attrs` group.  This would cause a slow down in the code and the effect would be proportional to the number of datasets that are contained within the yaml file.   Because of this I favor @kmpaul 's solution slightly more, but what do you both think?

Reply:
I'm going to see if I can rework my logic slightly in order to get @andersy005 to work more efficiently.

Reply:
@andersy005 and @kmpaul I have this working per @andersy005 suggestion.